README.md
# Makeiso Kuroko

[![996.icu](https://img.shields.io/badge/link-996.icu-red.svg)](https://996.icu)

This is a tool for making CentOS-7.3.1611 ISO for your own purpose.

This only works for CentOS-7.3.1611.

## Screenshot

### In BIOS mode

![makeiso-kuroko screenshot](screenshot_bios.png "makeiso-kuroko screenshot")

### In UEFI mode

![makeiso-kuroko screenshot](screenshot_uefi.png "makeiso-kuroko screenshot")

## Requirement

- An avaliable Linux platform. CentOS 7.5.1804 or newer is recommended because it includes rsync 3.1.1+
- A CentOS-7-x86_64-Everything-1611.iso image
- genisoimage (A RPM package is included in Everything 1611 iso image)
- createrepo (A RPM package is included in Everything 1611 iso image)
- rsync 3.1.1+ (not included in Everything 1611 iso image. A version 3.1.2 of rsync RPM package is included in this repository)

## Usage

Recommend order:

1. Modify the GLOBAL VARIABLE if needed.
2. Copy addtional files to `PAYLOAD_PATH` with a install.sh as the installation script, which you want to install after the system installed. Please notice **NO INTERACTIVE OPERATION IS ALLOWED**.
3. Run makeiso.sh to generate iso file. (Superuser is required because going to mount iso)

### GLOBAL VARIABLE

```
# INPUT
CENTOS7_EVERYTHING_ISO='/tmp/mountpoint/samba/share/CentOS-7-x86_64-Everything-1611.iso'
CENTOS7_EVERYTHING_ISO_MOUNTPOINT='/tmp/mountpoint/CentOS7-Everything-1611/'
PAYLOAD_PATH='./payload_sample/'
CONFIGDIR='boot.template/develop/'

# OUTPUT
NAMEPREFIX='PAYLOAD'
OUTPUTFILEDIR='./'
VERSION='v1.0.0'
TIMEZONE='UTC'

# Auto generated variables
VOLUMENAME=$NAMEPREFIX'-'`date +'%Y%m%d%H%M%S'`'-'$VERSION
VOLUMENAME_LABEL=`expr substr ${VOLUMENAME} 1 16`
FINALNAME=${VOLUMENAME}.iso
```

- **CENTOS7_EVERYTHING_ISO** Must be a accessiable CentOS-7-x86_64-Everything-1611.iso files.
- **PAYLOAD_PATH** is the addtional files you want to install after the system installation. After the system installation and auto reboot, `bash install.sh` will execute automatically once.
- **VOLUMENAME_LABEL** is for Volume ID and it only supports as long as 16 chars.

### Usage

```
Usage: # ./makeiso.sh -d [DEST_DIR=./] -v [RELEASE_VERSION=v1.0.0] -s [PAYLOAD_PATH=/root/payload_sample/] -7 [CENTOS7_EVERYTHING_ISO=/root/iso/CentOS-7-x86_64-Everything-1611.iso] -z [TIMEZONE=UTC]
```

Example 1:

`# ./makeiso.sh -7 /root/cifs/CentOS/CentOS-7-x86_64-Everything-1611.iso`

Example 2:

`# ./makeiso.sh -d /root/ -v test20200331 -s ./payload_sample/ -7 /root/cifs/CentOS/CentOS-7-x86_64-Everything-1611/CentOS-7-x86_64-Everything-1611.iso -z 'Asia/Shanghai'`

And you will get a ISO file. The default root password is 'makeiso-kuroko'.

**To set the root password, please refer to [rootpw - Set Root Password](https://docs.fedoraproject.org/en-US/Fedora/html/Installation_Guide/sect-kickstart-commands-rootpw.html)**

**Notice that all the network interfaces DHCP are enabled in kickstart-post-script. Your network configuration in the Install Guide won't work.**

**There is no config about installation destination in kickstart-post. You still need to select the installation destination**

* * *

## Project Details

### Packager

```
.
├── 83b61f9495b5f728989499479e928e09851199a8846ea37ce008a3eb79ad84a0-c7-minimal-x86_64-comps.xml
├── boot.template
│   └── develop
│       ├── EFI
│       │   └── BOOT
│       │       ├── grub.cfg
│       │       └── x86_64-efi
│       │           └── gfxterm_background.mod
│       └── isolinux
│           ├── isolinux.cfg
│           ├── payload-develop.cfg
│           └── splash.png
├── d918936f5019be3fb66e9981a28cb2a41477a2963d741d454f79377a22214f43-c7-x86_64-comps.xml
├── filelist
│   ├── centos_dvd_frame.list
│   └── minimal.list
├── generatefilelist.sh
├── makeiso.sh
├── payload_sample
│   └── install.sh
└── rsync-3.1.2-5.fc26.x86_64.rpm
```

- payload-develop.cfg, Kickstart files, including how to install automatically, the system language, network, root password, etc, A script which will executing after system installation and reboot is also included.
- centos_dvd_frame.list , CentOS-7-1611 DVD frameware files.
- minimal.list , RPM Package needed from Everything-DVD
- generatefilelist.sh , Diff 'installed RPM Package' between the current environment and Everything-DVD. This can be used to generate `your_rpm_payload.list`. **Notice that there is difference between MBR as EFI.**

### ISO

```
.
├── CentOS_BuildTag
├── EFI
│   ├── BOOT
│   │   ├── BOOTX64.EFI
│   │   ├── fonts
│   │   │   ├── TRANS.TBL
│   │   │   └── unicode.pf2
│   │   ├── grub.cfg
│   │   ├── grubx64.efi
│   │   ├── MokManager.efi
│   │   └── TRANS.TBL
│   └── TRANS.TBL
├── EULA
├── GPL
├── images
│   ├── efiboot.img
│   ├── pxeboot
│   │   ├── initrd.img
│   │   ├── TRANS.TBL
│   │   └── vmlinuz
│   └── TRANS.TBL
├── isolinux
│   ├── boot.cat
│   ├── boot.msg
│   ├── grub.conf
│   ├── initrd.img
│   ├── isolinux.bin
│   ├── isolinux.cfg
│   ├── memtest
│   ├── payload-develop.cfg
│   ├── splash.png
│   ├── TRANS.TBL
│   ├── vesamenu.c32
│   └── vmlinuz
├── LiveOS
│   ├── squashfs.img
│   └── TRANS.TBL
├── Packages
│   ├── *.rpm
├── PAYLOAD
│   ├── install.sh
│   ├── *
├── repodata
│   ├── *
│   ├── repomd.xml
│   └── TRANS.TBL
├── RPM-GPG-KEY-CentOS-7
├── RPM-GPG-KEY-CentOS-Testing-7
└── TRANS.TBL
```

### Destination platform root
```
.
├── anaconda-ks.cfg
├── initial-setup-ks.cfg
├── initstart_flag.log
├── initstart.log
├── initstart.sh
├── kickstart-post.log
├── original-ks.cfg
├── PAYLOAD
│   ├── install.sh
│   └── *
├── payload_install.log
├── payload_log_sample.log
└── rpm_qa.list
```

- anaconda-ks.cfg, The kickstart file for this installation at this time.
- initial-setup-ks.cfg, The kickstart file for this installation at this time if you installed Desktop.
- kickstart-post.log The log of the script in POST of kickstart.
- initstart.sh, the script which will be executed after the system installation and reboot, including 'disable YUM source', 'install PAYLOAD', and remove itself from `/etc/rc.d/rc.local`
- initstart_flag.log, the result of executing `initstart.sh`
- initstart.log, the log of `initstart.sh`
- PAYLOAD, Your PAYLOAD
- payload_install.log, the log of PAYLOAD/install.sh
- payload_log_sample.log，the output of PAYLOAD/install.sh if you used the default payload_sample
- rpm_qa_develop.list, the result of 'rpm -qa' when the whole installation is finished.

## Build your own ISO

Anything doesn't relate to the CentOS-7-x86_64-Everything ISO is supposed to be put in PAYLOAD and be installed by install.sh

For something relates to the ISO, you can install a new CentOS7 by yourself, and run generatefilelist.sh to generate new rpm filelist, and upgrade kickstart.cfg and comps.xml by yourself. (Or just edit it in the Payload install.sh. It's fine)

This project includes a comps.xml `d918936f5019be3fb66e9981a28cb2a41477a2963d741d454f79377a22214f43-c7-x86_64-comps.xml`, as same as CentOS-7-x86_64-Everything-1611. You should build your own comps.xml and kickstart.cfg for your project.

This project is just for helping developers who don't familiar how to buid a CentOS ISO. You should write Kickstart-file (payload-develop.cfg) for your own project purpose.

## Other things


### Idea about naming this 'make linux iso project' Kuroko

This one only works for centos7-1611. The year of 1611 is the year Galileo made naked-eye and telescopic studies of 'sunspots'. Sunspots is called '太阳黑子' in Chinese. '黑子' is also a part of name 'Shirai Kuroko'.

### Idea about naming the directory Payload

I folk project to another team of my company ,and named it 'Payload'. This word means a container, and in game Overwatch, it means vehicle to be escorted. The main idea is this directory is used for containing 3rd-party code.

Butt weight, The word 'Payload' in Team Fortress 2 also means BOMB! Because after I read the code and the project of the other team, I thought those would be a BOMB, it would explode at anytime. ["Bomb is friend! Come, visit friend!"](https://wiki.teamfortress.com/w/images/2/2b/Heavy_cartstaycloseoffense06.wav)

### Todo list

- Print logs not only in log file, but also in tty
- Replace rc.local using Systemd

## References

- [GRUB 2 Custom Splash Screen on RHEL 7 UEFI and Legacy ISO Image](http://www.tuxfixer.com/set-grub2-custom-splash-screen-on-rhel-7-centos-7-uefi-and-legacy-bios-iso-image/)
- [Grub2/Displays](https://help.ubuntu.com/community/Grub2/Displays#Troubleshooting_Splash_Images)
- [rootpw - Set Root Password](https://docs.fedoraproject.org/en-US/Fedora/html/Installation_Guide/sect-kickstart-commands-rootpw.html)

## Contribution

Any contributions are welcome. Pull request to branch **dev** please.

## License

Makeiso-Kuroko is licensed under Anti-996-License, which is adapted from the MIT license.

The default splash image is from ["Toaru Kagaku no Railgun - Shirai Kuroko" by Krukmeister](https://krukmeister.deviantart.com/art/Toaru-Kagaku-no-Railgun-Shirai-Kuroko-369206997) and it's licensed under a
[Creative Commons Attribution-Noncommercial-No Derivative Works 3.0 License](http://creativecommons.org/licenses/by-nc-nd/3.0/).

If you want to use this project for **commercial purposes**, you should **remove** or **replace** the default splash image.


CHANGELOG.md
# 1.3.1 2020-04-20

Not much critical changes since 1.3.0.

- Add new feature NAMEPREFIX
- Better color scheme in GRUB
- Better payload_sample/install.sh
- Minor bug fixes

# 1.3.0 2020-04-01

A lot of things have changed since 1.2.1

- The default root password is changed to 'makeiso-kuroko' (In case lazybones use root/root and make vulnerabilities)
- Grub timeout is 5 seconds now, both on BIOS mode and EFI mode
- Add a payload sample for users to understand how it works
- New splash screen image
- Add many environment checking before trying to make iso
- Add many prompting messages when making iso
- Change License to Anti-996-License
- Minor bug fixes

# 1.2.1  2018-06-16

- Minor adjust.

# 1.2 2018-06-03

- Now boot splash screen is supported. Both Legacy(MBR) and EFI are supported.
- Add ARG $CENTOS7_EVERYTHING_ISO path. Now not have to edit the GLOBAL VARIABLE of $CENTOS7_EVERYTHING_ISO.
- Add ARG $TIMEZONE. Now not have to edit the kickstart file to set Time Zone.
- Minor bug fixes.

# 1.1 2018-05-21

- Now EFI support is conformed.
- Delete a lot unnecessary file-copy in `minimal.list`. I have no idea why and how I added those unused files into filelist. Even the file is in the iso, they wouldn't be installed because they are not included in the comps.xml or kickstart. I just have no idea how did I make this mistake. For now the output iso file will be thinner (reduced about 20 MB).

# 1.0.1 2018-05-17

- Fix file name reference error.

# 1.0.0 2018-05-16

- Initial release


generatefilelist.sh
#!/bin/bash

set -euo pipefail
PACKAGE_PATH=/tmp/mountpoint/CentOS7-Everything/Packages
COUNT=0
FOUND=0
NOT_FOUND=0
FILELIST='filelist/your_rpm_payload.list'
if [[ -s $FILELIST ]]; then
    FILELIST_NEW='filelist/your_rpm_payload-'`date +'%Y%m%d%H%M%S'`'.list'
    echo "Notice: $FILELIST already exist. Your RPM payload list will be generated in $FILELIST_NEW"
    FILELIST=$FILELIST_NEW
fi
touch $FILELIST
FILELIST_TEMP=`mktemp`
rpm -qa | sort > $FILELIST_TEMP
#echo 'temp file: '$FILELIST_TEMP
echo ''

while read line
    do
        ((COUNT=COUNT+1))
        if [[ -f $PACKAGE_PATH/$line.rpm ]]; then
            #echo 'Found '$line.rpm
            echo $line.rpm >> $FILELIST
            ((FOUND=FOUND+1))
        else
            ((NOT_FOUND=NOT_FOUND+1))
            echo 'Not found: '$line.rpm
        fi
    done < $FILELIST_TEMP

echo 'Count: '$COUNT
echo 'Hit: '$FOUND
echo 'Miss: '$NOT_FOUND
echo ''
echo 'New filelist is generated to: '$FILELIST


payload_sample/install.sh
#!/bin/bash

set -euo pipefail
STARTTIME=`date +%s`

echo "Payload install.sh start at: "$STARTTIME | tee -a /root/payload_log_sample.log
#sleep 10
echo "I am a payload sample." | tee -a /root/payload_log_sample.log
#sleep 10

## For CentOS
echo -n "Your system: *** " | tee -a /root/payload_log_sample.log
echo -n "$(cat /etc/system-release)" | tee -a /root/payload_log_sample.log
echo " *** is ready." | tee -a /root/payload_log_sample.log

## For Ubuntu
#echo "****************************************** " | tee -a /root/payload_log_sample.log
#echo "Your lsb_release -a:" | tee -a /root/payload_log_sample.log
#lsb_release -a | tee -a /root/payload_log_sample.log
#echo "****************************************** " | tee -a /root/payload_log_sample.log

#sleep 10
echo "This line goes into payload_install.log"
#sleep 10
echo "This line goes into specified log file" >> /root/payload_log_sample.log
#sleep 10
echo "This line goes into both payload_install.log and specified log file" | tee -a /root/payload_log_sample.log
#sleep 10
ENDTIME=`date +%s`
SPENTTIME=$(($ENDTIME-$STARTTIME))
echo "Payload install.sh end at: "$ENDTIME". Time usage: "$SPENTTIME" seconds" | tee -a /root/payload_log_sample.log


makeiso.sh
#!/bin/bash

set -euo pipefail
STARTTIME=`date +%s`

cd $(dirname $(readlink -f $0))

# INPUT
CENTOS7_EVERYTHING_ISO='/tmp/mountpoint/samba/share/CentOS-7-x86_64-Everything-1611.iso'
CENTOS7_EVERYTHING_ISO_MOUNTPOINT='/tmp/mountpoint/CentOS7-Everything-1611/'
PAYLOAD_PATH='./payload_sample/'
CONFIGDIR='boot.template/develop/'

# OUTPUT
NAMEPREFIX='PAYLOAD'
OUTPUTFILEDIR='./'
VERSION='v1.0.0'
TIMEZONE='UTC'

# Global variables
RED='\033[0;31m'
GREEN='\033[0;32m'
NC='\033[0m'
ISOTMP='./iso_tmp'

usage()
{
    echo "Usage: $0 -n [NAMEPREFIX=$NAMEPREFIX] -d [DEST_DIR=$OUTPUTFILEDIR] -v [RELEASE_VERSION=$VERSION] -s [PAYLOAD_PATH=$PAYLOAD_PATH] -7 [CENTOS7_EVERYTHING_ISO=$CENTOS7_EVERYTHING_ISO] -z [TIMEZONE=$TIMEZONE]" >&2
    exit 1
}

print_horizontal_line()
{
    printf '%*s\n' "${COLUMNS:-$(tput cols)}" '' | tr ' ' -
}

while getopts 'n:d:v:s:7:z:h' arg
do
    case $arg in
        n)
            NAMEPREFIX=$OPTARG
            ;;
        d)
            OUTPUTFILEDIR=$OPTARG
            ;;
        v)
            VERSION=$OPTARG
            ;;
        s)
            PAYLOAD_PATH=$OPTARG
            ;;
        7)
            CENTOS7_EVERYTHING_ISO=$OPTARG
            ;;
        t)
            ISOTMP=$OPTARG
            ;;
        z)
            TIMEZONE=$OPTARG
            ;;
        h)
            usage
            ;;
        *)  #unkonw argument?
            echo -e $RED'unkonw argument'$NC
            usage
            ;;
    esac
done

# Auto generated variables
VOLUMENAME=$NAMEPREFIX'-'`date +'%Y%m%d%H%M%S'`'-'$VERSION
VOLUMENAME_LABEL=`expr substr ${VOLUMENAME} 1 16`
FINALNAME=${VOLUMENAME}.iso

# Before start. Check environment
echo ''
command -v rsync >/dev/null 2>&1 || { echo -e >&2 $RED'rsync is not installed.  Aborting.'$NC; exit 1; }
command rsync -n --info=progress2 makeiso.sh /dev/null >/dev/null 2>&1 || { echo -e >&2 $RED'This version of rsync is not supported. Version 3.1.1+ is recommended. Please upgrade.  Aborting.'$NC; exit 1; }
command -v createrepo >/dev/null 2>&1 || { echo -e >&2 $RED'createrepo is not installed.  Aborting.'$NC; exit 1; }
command -v genisoimage >/dev/null 2>&1 || { echo -e >&2 $RED'genisoimage is not installed.  Aborting.'$NC; exit 1; }

# Start
mkdir -p ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/
mountpoint ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/ >/dev/null ||
if [[ !`mountpoint ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/ >/dev/null` ]]; then
    echo 'Try mount '${CENTOS7_EVERYTHING_ISO} && mount -t iso9660 -o loop ${CENTOS7_EVERYTHING_ISO} ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/
fi

if [[ ! -d $PAYLOAD_PATH || ! -f  $PAYLOAD_PATH/install.sh ]]; then
    echo -e $RED'Payload not found in '$PAYLOAD_PATH'. Exit.'$NC
    exit 1
fi

print_horizontal_line
echo -e $GREEN Start to generate $FINALNAME at  `date -d "1970-01-01 UTC $STARTTIME seconds" +'%Y-%m-%d %T %z'` $NC
print_horizontal_line
echo 'Clear ISO temporary files at '${ISOTMP}'.'
rm -rf ${ISOTMP}
echo 'Copy ISO framework'
mkdir -p ${ISOTMP}/Packages
mkdir -p ${ISOTMP}/repodata
mkdir -p ${ISOTMP}/PAYLOAD
rsync -au --info=progress2 ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/isolinux iso_tmp/
xargs -n 1 -i cp -rs ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/{} ${ISOTMP} < ./filelist/centos_dvd_frame.list

echo -e 'Copy(Actually it'"'"'s "make symbolic links") minimal packages from CentOS7 DVD'
xargs -n 1 -i cp -s ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/Packages/{} ${ISOTMP}/Packages < ./filelist/minimal.list

echo 'Copy rsync-3.1.2-5.fc26.x86_64.rpm'
rsync -a --info=progress2 rsync-3.1.2-5.fc26.x86_64.rpm ${ISOTMP}/Packages

echo 'Copy PAYLOAD'
rsync -a --info=progress2 ${PAYLOAD_PATH} ${ISOTMP}/PAYLOAD

echo 'Copy Boot files'
rsync -a --info=progress2 $CONFIGDIR/ ${ISOTMP}/
sed -i 's/{$TITLE}/'$VOLUMENAME'/' ${ISOTMP}/isolinux/isolinux.cfg
sed -i 's/{$TITLE}/'$VOLUMENAME'/' ${ISOTMP}/EFI/BOOT/grub.cfg
#sed -i 's/{$VOLUMENAME_LABEL}/'$VOLUMENAME_LABEL'/' ${ISOTMP}/isolinux/isolinux.cfg
sed -i 's/{$VOLUMENAME_LABEL}/'$VOLUMENAME_LABEL'/' ${ISOTMP}/EFI/BOOT/grub.cfg
escaped_x="$(sed -e 's/[\/&]/\\&/g' <<< "$TIMEZONE")"
ESCAPED_TIMEZONE="${escaped_x}"
sed -i 's/{$TIMEZONE}/'$ESCAPED_TIMEZONE'/' ${ISOTMP}/isolinux/payload-develop.cfg

echo 'Create CentOS7 comps.xml'
rsync -a --info=progress2 83b61f9495b5f728989499479e928e09851199a8846ea37ce008a3eb79ad84a0-c7-minimal-x86_64-comps.xml ${ISOTMP}/repodata
createrepo -g repodata/83b61f9495b5f728989499479e928e09851199a8846ea37ce008a3eb79ad84a0-c7-minimal-x86_64-comps.xml ${ISOTMP}

echo ''
echo Generating ISO file in $OUTPUTFILEDIR/$FINALNAME
echo Version：$VERSION
echo Volume name：$VOLUMENAME
echo Volume label: $VOLUMENAME_LABEL

mkdir -p $OUTPUTFILEDIR
genisoimage \
    -quiet \
    -follow-links \
    -cache-inodes \
    -joliet-long \
    -R -J -T \
    -c isolinux/boot.cat \
    -b isolinux/isolinux.bin \
    -no-emul-boot \
    -boot-load-size 4 \
    -boot-info-table \
    -eltorito-alt-boot \
    -b images/efiboot.img \
    -no-emul-boot \
    -input-charset utf-8 \
    -V $VOLUMENAME_LABEL \
    -o $OUTPUTFILEDIR/$FINALNAME \
    ${ISOTMP}

echo "ISO is generated. Umount ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/"
# umount doesn't matter
umount ${CENTOS7_EVERYTHING_ISO_MOUNTPOINT}/ ||

echo ''
print_horizontal_line
echo -e $GREEN'ISO is generated at: '$NC $(realpath $OUTPUTFILEDIR/$FINALNAME)
ls -lh --color=auto $OUTPUTFILEDIR/$FINALNAME
ENDTIME=`date +%s`
SPENTTIME=$(($ENDTIME-$STARTTIME))
echo -e $GREEN'Finished. Time usage: '$SPENTTIME' seconds'$NC
print_horizontal_line

#clean_up_tips
echo 'ISO temporary files is located at '${ISOTMP}'. You may want to remove it.'
echo ''
